c7n/resources/iam.py

```diff
@@ -1108,83 +1108,118 @@ class UserDelete(BaseAction):
                 UserName=r['UserName'], GroupName=user_group['GroupName'])
 
-        # SSH Keys
+    def delete_ssh_keys(self, client, r):
         response = client.list_ssh_public_keys(UserName=r['UserName'])
         for key in response.get('SSHPublicKeys', ()):
             client.delete_ssh_public_key(
                 UserName=r['UserName'], SSHPublicKeyId=key['SSHPublicKeyId'])
 
-        # Signing Certificates
+    def delete_signing_certificates(self, client, r):
         response = client.list_signing_certificates(UserName=r['UserName'])
         for cert in response.get('Certificates', ()):
             client.delete_signing_certificate(
                 UserName=r['UserName'], CertificateId=cert['CertificateId'])
 
+    def delete_service_specific_credentials(self, client, r):
         # Service specific user credentials (codecommit)
         response = client.list_service_specific_credentials(UserName=r['UserName'])
         for screds in response.get('ServiceSpecificCredentials', ()):
             client.delete_service_specific_credential(
                 UserName=r['UserName'],
                 ServiceSpecificCredentialId=screds['ServiceSpecificCredentialId'])
 
+    def delete_user(self, client, r):
         client.delete_user(UserName=r['UserName'])
 
+    OPTIONS = {
+        'console-access': delete_console_access,
+        'access-keys': delete_access_keys,
+        'user-policies': delete_user_policies,
+        'mfa-devices': delete_hw_mfa_devices,
+        'groups': delete_groups,
+        'ssh-keys': delete_ssh_keys,
+        'signing-certificates': delete_signing_certificates,
+        'services-specific-credentials': delete_service_specific_credentials,
+        'user': delete_user,
+    }
+
+    schema = type_schema(
+        'delete',
+        options={
+            'type': 'array',
+            'items': {
+                'type': 'string',
+                'enum': list(OPTIONS.keys()),
+            }
+        })
+
+    permissions = (
+        'iam:ListAttachedUserPolicies',
+        'iam:ListAccessKeys',
+        'iam:ListGroupsForUser',
+        'iam:ListMFADevices',
+        'iam:ListServiceSpecificCredentials',
+        'iam:ListSigningCertificates',
+        'iam:ListSSHPublicKeys',
+        'iam:DeactivateMFADevice',
+        'iam:DeleteAccessKey',
+        'iam:DeleteLoginProfile',
+        'iam:DeleteSigningCertificate',
+        'iam:DeleteSSHPublicKey',
+        'iam:DeleteUser',
+        'iam:DetachUserPolicy',
+        'iam:RemoveUserFromGroup')
+
+    def process(self, resources):
+        client = local_session(self.manager.session_factory).client('iam')
+        for r in resources:
+            self.log.debug('Deleting user %s options: %s' %
+                (r['UserName'], self.data.get('options', 'all')))
+            self.process_user(client, r)
+
+    def process_user(self, client, r):
+        options = self.data.get('options', list(self.OPTIONS.keys()))
```

Comment:
in python2 dictionary ordering is random, so you could get the remove user first which would error out due to dependencies, the options dict needs to be an ordered dict to handle the order. it looks like your already handling this by hardcoding the 'user' option and then removing it separately, but using an ordered dict would resolve this more simply.